contracts/SpokePool.sol

```diff
@@ -1373,24 +1397,41 @@ abstract contract SpokePool is
         uint32 leafId,
         bytes32 l2TokenAddress,
         bytes32[] memory refundAddresses
-    ) internal {
+    ) internal returns (bool deferredRefunds) {
         if (refundAddresses.length != refundAmounts.length) revert InvalidMerkleLeaf();
 
-        address l2TokenAddressParsed = l2TokenAddress.toAddress();
+        uint256 spokeStartBalance = IERC20Upgradeable(l2TokenAddress.toAddress()).balanceOf(address(this));
+        uint256 totalRefundedAmount = 0;
 
         // Send each relayer refund address the associated refundAmount for the L2 token address.
         // Note: Even if the L2 token is not enabled on this spoke pool, we should still refund relayers.
-        uint256 length = refundAmounts.length;
-        for (uint256 i = 0; i < length; ++i) {
-            uint256 amount = refundAmounts[i];
-            if (amount > 0)
-                IERC20Upgradeable(l2TokenAddressParsed).safeTransfer(refundAddresses[i].toAddress(), amount);
+        for (uint256 i = 0; i < refundAmounts.length; ++i) {
+            totalRefundedAmount += refundAmounts[i];
+
+            if (totalRefundedAmount > spokeStartBalance) revert InsufficientSpokePoolBalanceToExecuteLeaf();
+            if (refundAmounts[i] > 0) {
+                bool success;
+                try
+                    IERC20Upgradeable(l2TokenAddress.toAddress()).transfer(
+                        refundAddresses[i].toAddress(),
+                        refundAmounts[i]
+                    )
+                returns (bool result) {
+                    success = result;
+                } catch {
+                    success = false;
+                }
+
+                if (!success) {
+                    relayerRefund[l2TokenAddress.toAddress()][refundAddresses[i].toAddress()] += refundAmounts[i];
+                    deferredRefunds = true;
+                }
+            }
         }
-
         // If leaf's amountToReturn is positive, then send L2 --> L1 message to bridge tokens back via
         // chain-specific bridging method.
         if (amountToReturn > 0) {
-            _bridgeTokensToHubPool(amountToReturn, l2TokenAddressParsed);
+            _bridgeTokensToHubPool(amountToReturn, l2TokenAddress.toAddress());
```

Comment:
We don't need to check here that `amountToReturn <= erc20.balanceOf(address(this))` because this case would be handled by the revert?